### ISSUE TITLE
[Fargate] [request]: Allow privileged mode

### ISSUE THREAD

**Issue by biltongza** [aws/containers-roadmap#1000]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
I want to be able to mount filesystems within my containers, and for that I need to be able to use privileged mode or add capabilities to the docker container.

**Which service(s) is this request for?**
Fargate

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
Specifically, I am trying to mount an overlay in a container. This requires the container be run in privileged mode.

**Are you currently working around this issue?**
No, there does not appear to be a workaround.

**Comment by ecliptik** [comment]
We are interested in this since we are using [Gitlab Kubernetes Runners](https://docs.gitlab.com/runner/executors/kubernetes.html) for our CI/CD pipelines and using Docker-in-Docker to build application containers and push them to ECR.

Currently we have larger ec2 nodes that are tainted to only run Gitlab runner pods to avoid resource contention with other applications/service on the EKS clusters. These nodes are only used when CI/CD jobs are in us, and are idle for periods of time. Autoscaling them is also a challenge when many jobs are running at once since cold-starts of EC2 instances can take 2-3 minutes to join the cluster.

By using [Fargate Profiles](https://docs.aws.amazon.com/eks/latest/userguide/fargate-profile.html), we would like to setup the `gitlab` namespace so that any job pods created there will run on their own Fargate node instead of a static ec2 node. This would allow us greater resource flexibility, cost optimization, and separation of jobs to avoid resource contention.

We can do some of our jobs in Fargate like this, but a large majority of our jobs are using `docker` and `docker-compose` to build images and run CI/CD tests and would require privileged mode and volume sharing of the Docker socket.

**Comment by wangyun1517** [comment]
All the steps are dockerized in my company, so there is no way to use fargate as the CI/CD agent.
We are really looking forward to have this ability to run docker in docker on the Fargate platform.

**Comment by Waples** [comment]
Bump so this not being forgotten. I need to get priv mode working on our current AWS EKS Fargate CI/CD GitLab PoC.

**Comment by sandan** [comment]
+1 For EKS Fargate. We'd like more capabilities to be supported (specifically SYS_PTRACE, DAC_READ_SEARCH) to a container's securityContext. So far it looks like these work:
```
        securityContext:
          capabilities:
            drop:
              - all
            add:
              - SETPCAP
              - SETUID
              - SETGID
```
There's an open request for SYS_PTRACE: https://github.com/aws/containers-roadmap/issues/1102.

**Comment by jonrau1** [comment]
+1 - Open Distro requires Privileged containers to run on EKS.

**Comment by FANMixco** [comment]
s3fs requires privileged also to sync folders.

**Comment by BinaryNate** [comment]
+1 for supporting `privileged` or `dockerSecurityOptions `. I want to run Chromium in Fargate, but `privileged` or `dockerSecurityOptions` is required to enable Chromium's app sandbox.

**Comment by amunhoz** [comment]
+1

**Comment by bradynotarize** [comment]
Required for any container using s3fs-fuse.

**Comment by andymac4182** [comment]
Looking at running OpenVSCode in Fargate but need this to be able to run containers inside so docker-in-docker. https://github.com/gitpod-io/openvscode-server

**Comment by mayurvin** [comment]
I need the privileged mode on fargate with ECS so that I can install and run a docker daemon. At this time, docker daemon cannot be started unless the task is being run with the --privileged mode.
Please add this capability as soon as possible.

**Comment by nielsvanoosterom-varias** [comment]
+1 for this.

**Comment by jarredkenny** [comment]
I would like to be able to use docker-forticlient as a base image for a Fargate task so that my task can reach private resources behind a VPN without a 24/7 IPSec tunnel on a VPC.

https://github.com/HybirdCorp/docker-forticlient

**Comment by askkhan84** [comment]
Need this for DinD to run Github self hosted runners on Fargate

**Comment by tekknolagi** [comment]
+1 to use nsjail inside Fargate

**Comment by ecliptik** [comment]
Any updates on this? I've tried using [buildah](https://buildah.io/), [podman](https://podman.io/), [rootless](https://rootlesscontaine.rs/), and [nerdctl](https://github.com/containerd/nerdctl/blob/master/docs/rootless.md) on fargate but none of them work.

Alternative is to bring up a separate docker build host and expose it over TCP, but that has major operational and security concerns.

**Comment by mreferre** [comment]
> Any updates on this? I've tried using [buildah](https://buildah.io/), [podman](https://podman.io/), [rootless](https://rootlesscontaine.rs/), and [nerdctl](https://github.com/containerd/nerdctl/blob/master/docs/rootless.md) on fargate but none of them work.
> 
> Alternative is to bring up a separate docker build host and expose it over TCP, but that has major operational and security concerns.

Can this help? Would Kaniko be a viable alternative for you? https://aws.amazon.com/blogs/containers/building-container-images-on-amazon-ecs-on-aws-fargate/

**Comment by pditommaso** [comment]
I think the main use case for this request is not to build containers with Fargate, but to be able to mount a file system via FUSE in fargate task. From the first comment:

> I want to be able to mount filesystems within my containers, and for that I need to be able to use privileged mode or add capabilities to the docker container.

**Comment by mreferre** [comment]
Fair enough. I am not sure I'd characterize that as the "main use case" but definitely it was the initial theme of this issue and we should respect that (even though people added additional reasons for needing privileged access, including builds, so it diluted a bit the initial intent).

**Comment by ecliptik** [comment]
> Would Kaniko be a viable alternative for you?

Kaniko is a no-go for us as it adds additional complexity into our Gitlab CI/CD pipelines.

> I think the main use case for this request is not to build containers with Fargate, but to be able to mount a file system via FUSE in fargate task. From the first comment:

Our main request is to build container images using Fargate. We've raised this with AWS Enterprise Support as well, specifically for running Gitlab Runners in Fargate and building  and pushing container images to ECR.

**Comment by anjalichaudhary** [comment]
> Looking at running OpenVSCode in Fargate but need this to be able to run containers inside so docker-in-docker. https://github.com/gitpod-io/openvscode-server


@andymac4182 Were you able to find a solution for this? If yes, could you please share your approach.

**Comment by dasanjaneyuludarla** [comment]
+1

**Comment by andymac4182** [comment]
@anjalichaudhary Ended up just using Gitpod online. Would of preferred Fargate but it doesn't work :(

**Comment by nicodp-leap** [comment]
> +1 for supporting `privileged` or `dockerSecurityOptions `. I want to run Chromium in Fargate, but `privileged` or `dockerSecurityOptions` is required to enable Chromium's app sandbox.

Did you find any solution? I'm having the same problem.

**Comment by binarynate** [comment]
@nicodp-leap I haven't found a solution, so I'm running Chromium on EC2 instead of Fargate. I would still love to be able to switch it over to Fargate, though.

**Comment by gugaiz** [comment]
@binarynate @nicodp-leap one option is to use AWS Batch with EC2 instances in the background. You must set min vCPU to 0 when configuring "compute environments". That will lunch an EC2 instance, run the job in privileged mode and finish the instance after that. You can even use EC2 spot instances for that.

**Comment by kerashanog** [comment]
+1 for option runtime like [Nestybox/sysbox](https://github.com/nestybox/sysbox) on Fargate

**Comment by jimmy-bergman-oneplatformer** [comment]
+1 for this, would like to be able to use nsjail from Fargate containers

**Comment by jimmybergman** [comment]
+1 for this, would like to be able to use nsjail from Fargate containers

**Comment by acidjazz** [comment]
+1 for this, need to use s3fuse

**Comment by FearlessHyena** [comment]
It's been almost 3 years since this was opened. Is it at least being considered?

**Comment by pluangashish** [comment]
Hi +1 for this -- many images wants to be run in privileged mode true or either tell us better solution for this

**Comment by davies** [comment]
+1 for this to use juicefs in containers, which turns S3 into a shared disk but requires FUSE.

**Comment by luthfimasruri** [comment]
Any solution for this?

**Comment by S0LERA** [comment]
+1 We need to use Rancher, that requires --privileged

**Comment by hilary-b** [comment]
+1 for building docker images

**Comment by junekhan** [comment]
+1 also for building docker images. We applied Kaniko as a makeshift, but now we need to build multi-arch docker images which Kaniko can't do.

**Comment by ClaytonOlleyNutrien** [comment]
Doesn't privileged mode basically mean root access to the host machine? On a cloud resource you don't "own" or manage directly? Maybe I'm missing something but I can't see how we'll ever get root level access to a Fargate machine.

**Comment by ClaytonOlleyNutrien** [comment]
Yeah I understand that but I think it's also often shared with workloads from other accounts/users so getting root access to it seems unlikely.

**Comment by esamattis** [comment]
Those who are interested in running jails or sandbox processes (ping @tekknolagi) you might not actually need privileged containers but just an ability to create user namespaces which normally do not require any extra permissions. Unprivileged user namespaces are the canonical way to create rootless sandboxes in Linux. See this issue for details: https://github.com/aws/containers-roadmap/issues/2102

**Comment by one-summers-day** [comment]
+1 For rust.. when rust invokes an error it uses usys/unix/time.rs and that requires privileged mode (sudo) to run correctly T_T

**Comment by DoobleD** [comment]
We're interested in this capability as well. In our case, we use iptables and ipset from our container to dynamically block offending IP addresses, and without the priviledged mode, we get these errors:

`Kernel error received: Operation not permitted`

Unless there is a (simple) solution to this, this is a deal breaker for us. We can't move to Fargate.

We also use `sysctl` to tune some kernel parameters from the container (`net.core.somaxconn`, `net.ipv4.ip_local_port_range`, `fs.inotify.max_user_instances`). Even though there is now solutions to set some of these in configuration, the priviledged mode would make it much simpler by allowing the use of `sysctl` from the container.

**Comment by fernandomullerjr** [comment]
+1 We need to use Rancher, that requires --privileged

**Comment by DoobleD** [comment]
Just curious whether we might get this ability someday, or does the AWS team knows already that it won’t happen?

**Comment by one-summers-day** [comment]
Most likely isn't happening. Apparently it's a security vulnerability of some sort

**Comment by BwL1289** [comment]
Coming up on 5 years in a few months...

**Comment by claytonolley** [comment]
> @ClaytonOlleyNutrien The host _machine_ in question for Fargate containers is a _virtual_ machine, not the underlying physical hardware that runs the VM.

You'll still get root access to the VM which could be running Fargate containers for who knows how many customers. I'm guessing they cannot or will not add VM tenancy per account to Fargate.

**Comment by dfuentes77** [comment]
This is would be very beneficial to the capability of Fargate.  I'm guessing this has already been considered but I would want to emphasize that if/when this is allowed, the idea of a hop limit configuration for Fargate tasks is now relevant.  I would want containers inside a Fargate container to be able to use the AWS permissions attached to the Fargate container.  I hope this is considered in the implementation.

**Comment by AbhishekNautiyal** [comment]
This feature has been delivered with the Amazon ECS Managed Instances launch as covered in https://github.com/aws/containers-roadmap/issues/2452.  Refer to our [documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/managed-instances-security.html) to learn more.

**Comment by AbhishekNautiyal** [comment]
We have delivered support for privileged containers and elevated privileges on fully-managed compute with the ECS Managed Instances launch (https://github.com/aws/containers-roadmap/issues/2452). ECS Managed Instances is a fully managed compute option that delivers the same operational and security benefits as Fargate while providing access to the full range of EC2 capabilities.

We are aware that this feature request was originally opened for Fargate. While Managed Instances enables privileged mode, some customers may still want this capability for Fargate compute. Given Fargate's security model, we currently do not have plans to introduce elevated privileges. However, we continually seek your feedback to help shape our roadmap. I am closing this issue to help us better track and prioritize feature requests specific to each compute option. Please open a new issue if you require privileged mode support on Fargate, including your use case(s), or if you have any other requests.

As always, we appreciate your engagement and feedback.
